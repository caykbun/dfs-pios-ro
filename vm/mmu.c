#include "rpi.h"
#include "rpi-constants.h"
#include "rpi-interrupts.h"
#include "libc/helper-macros.h"
#include "vm/mmu.h"

/***********************************************************************
 * the following code is given
 */

cp_asm(dac, p15, 0, c3, c0, 0)

void cp15_domain_ctrl_wr(uint32_t r);

enum { OneMB = 1024 * 1024 };

// maximum number of 1mb section entries for a 32-bit address space
enum { MAX_SEC_PTE = 4096};

void mmu_reset(void);
void mmu_enable_set_asm(cp15_ctrl_reg1_t c);
void mmu_disable_set_asm(cp15_ctrl_reg1_t c);
void mmu_sync_pte_mods(void);

// given.
int mmu_is_enabled(void) {
    return cp15_ctrl_reg1_rd().MMU_enabled != 0;
}

// given memory allocated at <addr> do page table initialization.
fld_t *mmu_pt_init(void *addr, unsigned nbytes) {
    fld_t *pt = addr;
    unsigned n = 4096 * sizeof *pt;
    demand(mod_pow2_ptr(addr, 14), must be 14-bit aligned!);
    demand(nbytes == n, only handing full allocation);
    memset(pt, 0, n);
    return pt;
}

// allocate page table with <n> entries.
fld_t *mmu_pt_alloc(unsigned n) {
    demand(n == 4096, we only handling a fully-populated page table right now);
    unsigned nbytes = 4096 * sizeof(fld_t);

    // first-level page table is 4096 entries.
    fld_t *pt = kmalloc_aligned(nbytes, 1<<14);

    demand(is_aligned_ptr(pt, 1<<14), must be 14-bit aligned!);
    return mmu_pt_init(pt, nbytes);
}


// disable the mmu by setting control register 1
// to <c:32>.
// 
// we use a C veneer over the assembly (mmu_disable_set_asm)
// so we can easily do assertions: the real work is 
// done by the asm code (you'll write this next time).
void mmu_disable_set(cp15_ctrl_reg1_t c) {
    assert(!c.MMU_enabled);
    assert(!c.C_unified_enable);
    // You'll implement this next lab.
    mmu_disable_set_asm(c);
}

// disable the MMU by flipping the enable bit.   we 
// use a C vener so we can do assertions and then call
// out to assembly to do the real work (you'll write this
// next time).
void mmu_disable(void) {
    cp15_ctrl_reg1_t c = cp15_ctrl_reg1_rd();
    assert(c.MMU_enabled);
    c.MMU_enabled=0;
    mmu_disable_set(c);
}

// enable the mmu by setting control reg 1 to
// <c>.   we start in C so we can do assertions
// and then call out to the assembly for the 
// real work (you'll write this code next time).
void mmu_enable_set(cp15_ctrl_reg1_t c) {
    assert(c.MMU_enabled);
    // you'll implement this next lab.
    mmu_enable_set_asm(c);
}

// enable mmu by flipping enable bit.
void mmu_enable(void) {
    cp15_ctrl_reg1_t c = cp15_ctrl_reg1_rd();
    assert(!c.MMU_enabled);
    c.MMU_enabled = 1;
    mmu_enable_set(c);
}

// C end of this: does sanity checking then calls asm.
void set_procid_ttbr0(unsigned pid, unsigned asid, fld_t *pt) {
    assert((pid >> 24) == 0);
    assert(pid > 64);
    assert(asid < 64 && asid);
    // you'll implement this next lab.
    cp15_set_procid_ttbr0(pid << 8 | asid, pt);
}

/**************************************************************************
 * for lab 12: implement the following code.
 */

// you'll need implement this: 
//   iterate over the page table <pt> setting up the mappings from
//   va->pa for <nsec> sections.  use domain id <dom>.
void mmu_map_sections(fld_t *pt, unsigned va, unsigned pa, unsigned nsec, uint32_t dom) {
    assert(dom < 16);
    assert(mod_pow2(va, 20));
    assert(mod_pow2(pa, 20));

    // implement
    for (unsigned i = 0; i < nsec; ++i) {
        mmu_map_section(pt, va, pa, dom);
        va += 1 << 20;
        pa += 1 << 20;
    }

    
}

// lookup va in pt and return the pte entry.
fld_t * mmu_lookup_section(fld_t *pt, unsigned va) {
    assert(mod_pow2(va, 20));
    fld_t *pte = 0;

    // implement
    // pte = staff_mmu_lookup_section(pt,va);
    pte = pt + (va >> 20);
    // pte = (fld_t *) addr;

    // for today: tag should be set.  in the future you'd return 0.
    demand(pte->tag, invalid section);
    return pte;
}

// set the ap bits for va,nsec to <perm>
void mmu_mprotect(fld_t *pt, unsigned va, unsigned nsec, unsigned perm) {
    demand(perm <= 0b11, invalid permission);

    // you need to implement this.
    // trace("%x\n", mmu_lookup_section(pt, va & ~((1 << 20) - 1))->AP);
    // staff_mmu_mprotect(pt,va,nsec, perm);
    // trace("%x\n", mmu_lookup_section(pt, va & ~((1 << 20) - 1))->AP);
    fld_t *pte = mmu_lookup_section(pt, va & ~((1 << 20) - 1));
    for (unsigned i = 0; i < nsec; ++i) {
        pte->AP = perm;
        pte++;
    }

    // must call this routine on each PTE modification (you'll implement
    // next lab).
    mmu_sync_pte_mods();
}

// set so that we use armv6 memory.
// this should be wrapped up neater.  broken down so can replace 
// one by one.
//  1. the fields are defined in vm.h.
//  2. specify armv6 (no subpages).
//  3. check that the coprocessor write succeeded.
void mmu_init(void) { 
    // reset the MMU state: you will implement next lab
    mmu_reset();

    // trivial: RMW the xp bit in control reg 1.
    // leave mmu disabled.
    // todo("read control reg 1, turn on XP bit (non-back compat)");
    struct control_reg1 temp = cp15_ctrl_reg1_rd();
    temp.XP_pt = 1;

    cp15_ctrl_reg1_wr(temp);

    // make sure write succeeded.
    struct control_reg1 c1 = cp15_ctrl_reg1_rd();
    assert(c1.XP_pt);
    assert(!c1.MMU_enabled);
}

// set f->sec_base_addr correctly: called by <mmu_map_section>
static void fld_set_base_addr(fld_t *f, unsigned addr) {
    // 20 b/c we have 1MB sections.
    demand(mod_pow2(addr,20), addr is not aligned!);

    // make sure if you read it back, it's what you set it to.
    f->sec_base_addr = addr >> 20;

    // if the previous code worked, this should always succeed.
    assert((f->sec_base_addr << 20) == addr);
}

// create a mapping for <va> to <pa> in the page table <pt>
// for now: 
//  - assume domain = 0.
//  - global mapping
//  - executable
//  - don't use APX
//  - TEX,A,B: for today mark everything as non-cacheable 6-15/6-16 in armv1176
fld_t * mmu_map_section(fld_t *pt, uint32_t va, uint32_t pa, uint32_t dom) {
    // only 16 domains.
    assert(dom < 16);
    // 20 b/c we have 1MB sections.
    assert(mod_pow2(va, 20));
    assert(mod_pow2(pa, 20));

    // assign pte: call <fld_set_base_addr> to set <sec_base_addr>
    // set each pte entry to: 
    //   1. global
    //   2. AP = full access in user and kernel (B4-9)
    //   3. APX = 0
    //   4. domain to <dom>
    //   5. TEX strongly ordered (B4-12)
    //   6. executable.
    // fld_t *pte = staff_mmu_map_section(pt, va, pa, dom);
    // trace("WARNING %x\n", pte);
    fld_t *pte = pt + (va >> 20);

    fld_set_base_addr(pte, pa);
    pte->nG = 0;
    pte->APX = 0;
    pte->AP = 0b11;
    pte->domain = dom;
    pte->TEX = 0b000;
    pte->XN = 0;
    pte->tag = 0b10;
    pte->B = 0;
    pte->C = 0;
    pte->IMP = 0;
    pte->super = 0;
    pte->_sbz1 = 0;
    pte->S = 0;
    // trace("WARNING %x\n", pte);

    fld_print(pte);
    printk("my.pte@ 0x%x = %b\n", pt, *(unsigned*)pte);
    hash_print("PTE crc:", pte, sizeof *pte);
    return pte;
}

// read and return the domain access control register
uint32_t domain_access_ctrl_get(void) {
    // return 0;
    return dac_get();
    // return staff_domain_access_ctrl_get();
}

// b4-42
// set domain access control register to <r>
// void domain_access_ctrl_set(uint32_t r) {
//     staff_domain_access_ctrl_set(r);
//     assert(domain_access_ctrl_get() == r);
// }

void domain_access_ctrl_set(uint32_t r) {
    // staff_cp15_domain_ctrl_wr(r);
    // dac_set(r);
    cp15_domain_ctrl_wr(r);
    // prefetch_flush();
}