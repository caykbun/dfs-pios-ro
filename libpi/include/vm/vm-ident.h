#ifndef __VM_IDENT_H__
#define __VM_IDENT_H__

#include "rpi.h"
#include "rpi-constants.h"
#include "rpi-interrupts.h"

#include "mmu.h"

fld_t * vm_ident_mmu_init(int start_p, void *int_vec);
void vm_ident_mmu_on(fld_t *pt);
void vm_ident_mmu_off(void);

// we use a non-zero domain id to test things out.
// (this shouldn't be a global) 
enum { dom_id = 1};

// kinda lame, but useful to have around.
enum { OneMB = 1024 * 1024 };

// used for tests.
extern volatile struct proc_state {
    fld_t *pt;
    unsigned sp_lowest_addr;
    unsigned fault_addr;
    unsigned dom_id;
    unsigned fault_count;

    // used for the die checks.
    unsigned die_addr;
} proc;



#endif
