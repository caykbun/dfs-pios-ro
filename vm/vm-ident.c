// What to do:
//   - replace the staff calls with calls to your routines in mmu.c
//
// section based: each section is 1MB.
//
// trivial setup of identity mappings where we map a virtual section 
// to the same physical section (e.g. 0x100000 maps to 0x100000)
//
// assumes a very simple address space:
//    - 1 section of code (starting at address 0x0)
//    - 1 section of user stack (ending at STACK_ADDR)
//    - 1 section of interrupt stack (ending at INT_STACK_ADDR)
//    - 1 section of heap (starting at one MB: 0x100000)
//    - 3 sections of GPIO ( 0x20000000 ---  0x20200000)
#include "vm/vm-ident.h"
#include "vector-base.h"

enum { MB = 1024 * 1024 };

// void mmu_reset(void);

// do one time initialization of MMU, setup identity mappings, and start
// running in mmu
fld_t * vm_ident_mmu_init(int start_p, void *int_vec) {
    // 1. init
    mmu_init();
    assert(!mmu_is_enabled());

    // ugly hack: we start by allocating a single page table at a known address.
    fld_t *pt = mmu_pt_alloc(4096);
    assert(pt == (void*)0x100000);

    // 2. setup mappings

    // map the first MB: shouldn't need more memory than this.
    mmu_map_section(pt, 0x0, 0x0, dom_id);
    // map the heap/page table: for lab cksums must be at 0x100000.
    mmu_map_section(pt, 0x100000,  0x100000, dom_id);

    mmu_map_section(pt, 0x20000000, 0x20000000, dom_id);
    mmu_map_section(pt, 0x20100000, 0x20100000, dom_id);
    mmu_map_section(pt, 0x20200000, 0x20200000, dom_id);
    
    mmu_map_section(pt, STACK_ADDR-MB, STACK_ADDR-MB, dom_id);
    mmu_map_section(pt, INT_STACK_ADDR-MB, INT_STACK_ADDR-MB, dom_id);

    // 3. install fault handler to catch if we make mistake.
    int_vec_reset(int_vec);

    // 4. start the context switch:

    // set up permissions for the one domain we use rn.
    domain_access_ctrl_set(0b01 << dom_id*2);

    if(start_p)
        vm_ident_mmu_on(pt);
    return pt;
}

void vm_ident_mmu_on(fld_t *pt) {
    // this code uses the sequence on B2-25
    set_procid_ttbr0(0x140e, dom_id, pt);

    // note: this routine has to flush I/D cache and TLB, BTB, prefetch buffer.
    mmu_enable();
    assert(mmu_is_enabled());
}

void vm_ident_mmu_off(void) {
    mmu_disable();
    assert(!mmu_is_enabled());
}
