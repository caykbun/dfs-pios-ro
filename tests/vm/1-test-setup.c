/* engler, 140e: sorta cleaned up test code from 14-vm.  */
#include "rpi.h"
#include "rpi-constants.h"
#include "rpi-interrupts.h"

#include "vm/vm-ident.h"
#include "vm/mmu.h"


#if 0
// for today, just crash and burn if we get a fault.
void data_abort_vector(unsigned lr) {
    unsigned fault_addr;
    // b4-44
    asm volatile("MRC p15, 0, %0, c6, c0, 0" : "=r" (fault_addr));
    staff_mmu_disable();
    panic("data_abort fault at %x\n", fault_addr);
}

// shouldn't happen: need to fix libpi so we don't have to do this.
void interrupt_vector(unsigned lr) {
    staff_mmu_disable();
    panic("impossible\n");
}
#endif

/*
 * an ugly, trivial test run.
 *  1. setup the virtual memory sections.
 *  2. enable the mmu.
 *  3. do some simple tests.
 *  4. turn the mmu off.
 */ 
void vm_test(void) {
    extern uint32_t interrupt_vector_vm[];

    fld_t *pt = vm_ident_mmu_init(1, (void *)interrupt_vector_vm);
    assert(mmu_is_enabled());

    /*********************************************************************
     * 5. trivial test: write a couple of values, make sure they succeed.
     *
     * you should write additional tests!
     */

    // read and write a few values.
    int x = 5, v0, v1;
    assert(get32(&x) == 5);

    v0 = get32(&x);
    printk("doing print with vm ON\n");
    x++;
    v1 = get32(&x);


    /*********************************************************************
     * done with trivial test, re-enable.
     */
    vm_ident_mmu_off();
    printk("OFF\n");

    // 7. make sure worked.
    assert(v0 == 5);
    assert(v1 == 6);
    printk("******** success ************\n");
}

void notmain() {
    kmalloc_init_set_start((void*)OneMB, OneMB);
    printk("implement one at a time.\n");
    check_vm_structs();
    vm_test();
}
