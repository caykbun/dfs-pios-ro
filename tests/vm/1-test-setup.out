HASH:PTE crc:: hash=0x2f782a81,nbytes=4
HASH:PTE crc:: hash=0x5f6614fe,nbytes=4
HASH:PTE crc:: hash=0x314dc8fc,nbytes=4
HASH:PTE crc:: hash=0x93a2dadc,nbytes=4
HASH:PTE crc:: hash=0xb76d0d60,nbytes=4
HASH:PTE crc:: hash=0xe22b84dd,nbytes=4
HASH:PTE crc:: hash=0x87a285bd,nbytes=4
